### PR TITLE
Support Multiple Keys for WebAuthn

### PR DESCRIPTION
**Pull request by developStorm**

### CODE REVIEW

File: app/dashboard/templates/dashboard/fido_setup.html [hunk 1]
```diff
@@ -52,7 +56,13 @@ <h1 class="h2 text-center">Register Your Security Key</h1>
         }
 
         $("#btnRegisterKey").click(registerKey);
-        $('document').ready(registerKey());
+
```
[the remark this answers is not in the document]
Comment (nguyenkims) [reply app/dashboard/templates/dashboard/fido_setup.html:59]: This has another advantage of enforcing `required` check on the key name.

File: app/dashboard/views/fido_setup.py [hunk 2]
```diff
@@ -1,37 +1,43 @@
...
-
     if not current_user.can_use_fido:
         flash(
             "This feature is currently in invitation-only beta. Please send us an email if you want to try",
             "warning",
         )
         return redirect(url_for("dashboard.index"))
 
+    if current_user.fido_uuid is not None:
+        fido_model = FIDO.filter_by(uuid=current_user.fido_uuid).all()
```
Comment (nguyenkims) [inline app/dashboard/views/fido_setup.py:37]: Probably a nitpicking here :), I think we can rename `fido_model` to `fidos`

File: app/dashboard/views/fido_setup.py [hunk 3]
```diff
@@ -91,6 +112,16 @@ def fido_setup():
     registration_dict = credential_create_options.registration_dict
     del registration_dict["extensions"]["webauthn.loc"]
 
+    # Prevent user from adding duplicated keys
+    for record in fido_model:
```
Comment (nguyenkims) [inline app/dashboard/views/fido_setup.py:116]: `record` can be renamed to `fido` here :)

**Comment by developStorm** [comment]
UUID can provide more entropy compared to userid (just a integer), and we can reset it in case. It doesn't cost more resources so I don't think there is a disadvantage for using uuid?

File: app/dashboard/templates/dashboard/fido_manage.html [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+            <th scope="col">Linked At</th>
+            <th scope="col" class="text-center">Operation</th>
+          </tr>
+        </thead>
+        <tbody>
+          {%for key in keys%} 
+          <tr>
+            <th scope="row">{{ key.id }}</th>
+            <td>{{ key.name }}</td>
+            <td><script>document.write(new Date('{{ key.created_at }}').toLocaleString());</script></td>
```
[the remark this answers is not in the document]
Comment (developStorm) [reply app/dashboard/templates/dashboard/fido_manage.html:36]: Will this trick adjust the time string with user's timezone? My purpose of put that in the front end is to make sure users can see their local time

File: app/dashboard/templates/dashboard/fido_manage.html [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+            <td class="text-center"><button class="btn btn-outline-danger" onclick="$('#credential_id').val('{{ key.credential_id }}'); $('#formManageKey').submit();">Unlink</button></td>
```
[the remark this answers is not in the document]
Comment (developStorm) [reply app/dashboard/templates/dashboard/fido_manage.html:37]: We can -- but that will required manually handle all form checks and CSRF. Is there any reason that you wish to use multiple forms? Cause I think it should have the same behavior as now.

[the remark this answers is not in the document]
Comment (developStorm) [reply app/dashboard/templates/dashboard/fido_manage.html:37]: We can -- but that will required manually handle all form checks and CSRF. Is there any reason that you wish to use multiple forms? Cause I think it should have the same behavior as now.

File: app/dashboard/views/enter_sudo.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        "dashboard/enter_sudo.html", password_check_form=password_check_form
+    )
+
+
+def sudo_required(f):
+    @wraps(f)
+    def wrap(*args, **kwargs):
+        # Reset sudo mode in every 60s under dev mode
+        # SUDO_GAP = 900 if not DEBUG else 60
+        SUDO_GAP = 900
```
[the remark this answers is not in the document]
Comment (developStorm) [reply app/dashboard/views/enter_sudo.py:50]: Hmm, at the beginning of this file?

File: app/dashboard/templates/dashboard/fido_setup.html (hunk 1 above)
[the remark this answers is not in the document]
Comment (developStorm) [reply app/dashboard/templates/dashboard/fido_setup.html]: https://github.com/simple-login/app/pull/199/commits/c0a751ff132432f15e7ef4d17b9d724b43a2d01f

File: app/dashboard/views/enter_sudo.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (developStorm) [reply app/dashboard/views/enter_sudo.py]: https://github.com/simple-login/app/pull/199/commits/5c3d2c19c8c384ca9c03b859370187a3d7893bc5

File: app/dashboard/views/fido_manage.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+class FidoManageForm(FlaskForm):
+    credential_id = HiddenField("credential_id", validators=[validators.DataRequired()])
+
+
+@dashboard_bp.route("/fido_manage", methods=["GET", "POST"])
+@login_required
+@sudo_required
+def fido_manage():
+    if not current_user.fido_enabled():
+        flash("You haven't registed a security key", "warning")
```
[the remark this answers is not in the document]
Comment (developStorm) [reply app/dashboard/views/fido_manage.py]: https://github.com/simple-login/app/pull/199/commits/7d1a744fe2478db00bc3e7123767c0d8e496d3ec

File: app/models.py [hunk 8]
```diff
@@ -120,6 +120,19 @@ class AliasGeneratorEnum(EnumE):
     uuid = 2  # aliases are generated based on uuid
 
 
+class FIDO(db.Model, ModelMixin):
```
[the remark this answers is not in the document]
Comment (developStorm) [reply app/models.py]: https://github.com/simple-login/app/pull/199/commits/ea914e0378040e034e1c423c6844c862be3d131f

File: app/dashboard/views/fido_setup.py (hunk 2 above)
Comment (developStorm) [reply app/dashboard/views/fido_setup.py]: https://github.com/simple-login/app/pull/199/commits/7bd97e13b008b4a3b265f44fc96c2cf8907b3b2a

File: app/dashboard/views/fido_setup.py (hunk 3 above)
Comment (developStorm) [reply app/dashboard/views/fido_setup.py]: https://github.com/simple-login/app/pull/199/commits/7bd97e13b008b4a3b265f44fc96c2cf8907b3b2a

File: app/dashboard/templates/dashboard/fido_manage.html (hunk 4 above)
[the remark this answers is not in the document]
Comment (developStorm) [reply app/dashboard/templates/dashboard/fido_manage.html]: https://github.com/simple-login/app/pull/199/commits/4fd7bf40abcc5a5a44febde18819d0a2da740887

File: app/dashboard/templates/dashboard/fido_manage.html (hunk 5 above)
[the remark this answers is not in the document]
Comment (nguyenkims) [reply app/dashboard/templates/dashboard/fido_manage.html:37]: I see, let's keep this approach then.

**Comment by nguyenkims** [comment]
> UUID can provide more entropy compared to userid (just a integer), and we can reset it in case. It doesn't cost more resources so I don't think there is a disadvantage for using uuid?

Hey I agree that `uuid` offers more flexibility (and is also more scalable) than `integer id`. However `integer id` column is currently  used to link tables together so using `uuid` to link `fido` and `user` is not consistent with existing code I think. 

Otherwise apart from this foreign key that I still feel not consistent with the existing code, the PR looks good to me!

**Comment by developStorm** [comment]
Hmm so let me make sure -- you still wish me to change that key to `id`, right?

**Comment by nguyenkims** [comment]
> Hmm so let me make sure -- you still wish me to change that key to `id`, right?

Yes that would be great!

**Comment by developStorm** [comment]
I double checked the related documents and it turns out use integer id is not recommended:

> ...Must be pseudo random...

https://developers.yubico.com/WebAuthn/WebAuthn_Developer_Guide/User_Handle.html

> ...It is RECOMMENDED to let the user handle be 64 random bytes, and store this value in the user’s account.

https://www.w3.org/TR/webauthn/#sctn-user-handle-privacy

**Comment by nguyenkims** [comment]
@developStorm Thanks for the info!

**Comment by developStorm** [comment]
Just found [this line](https://github.com/simple-login/app/blob/master/app/dashboard/views/fido_setup.py#L87) has a bug, should not have `query.` in it. 

I encountered a 500 error when testing in the actual SL env, but this was not shown up when I debug locally, weird. 😐

**Comment by nguyenkims** [comment]
> Just found [this line](https://github.com/simple-login/app/blob/master/app/dashboard/views/fido_setup.py#L87) has a bug, should not have `query.` in it.
> 
> I encountered a 500 error when testing in the actual SL env, but this was not shown up when I debug locally, weird. 😐
> 
> @nguyenkims

Actually the bug is here https://github.com/simple-login/app/commit/22c5513909b86235d3cf9d8bfff762cae3dc3b14

I'm going to deploy the fix as soon as it finishes building :).

The `filter_by` can be used with both `RecoveryCode` or `RecoveryCode.query` thanks to this mixin https://github.com/simple-login/app/blob/097ac771b05e1db25e09603af5774a3692aa67c1/app/models.py#L59

**Comment by nguyenkims** [comment]
@developStorm deployed :). I should have tested it on prod as well. On staging I strangely haven't encountered any issue ...

**Comment by developStorm** [comment]
Maybe is a behavior difference between SQLite and PostgreSQL?

**Comment by nguyenkims** [comment]
@developStorm I have no idea. I wonder if this can be due to my manual intervention of setting `fido_uuid` to `null` when other fido columns are removed from `users` table 🤔

**Comment by developStorm** [comment]
Isn't it already fixed now? Null is the default value and that should be fine...🧐

**Comment by nguyenkims** [comment]
Yes the fix has been deployed.

**Pull request merged**